File: lib/world/provider.rb
module Synthea
  class Provider
    cattr_accessor :provider_list # Array - all providers that are imported
    cattr_accessor :services # Hash of services to hospitals that provide them

    attr_accessor :attributes # Hash
    attr_accessor :utilization # Hash
    attr_accessor :services_provided # Array - services include ambulatory, emergency, and inpatient

    def initialize(properties, coordinates)
      @attributes = properties
      @attributes[:coordinates] = coordinates

      # labs = lipid_panel, basic_metabolic_panel, electrolytes_panel
      # diabetes_labs = lipid_panel
      @utilization = { encounters: 0, procedures: 0, labs: 0, prescriptions: 0 }

      Provider.provider_list.push(self)

      @services_provided = []
      services_split = properties['services_provided'].split
      services_split.each do |service|
        # add service to list of services provided by provider
        @services_provided.push(service.to_sym)

        # add provider to hash of services
        if Provider.services.key?(service.to_sym)
          Provider.services[service.to_sym].push(self)
        else
          Provider.services[service.to_sym] = [self]
        end
      end
    end

    # rubocop:disable Style/ClassVars
    # from module.rb
    def self.provider_list
      @@provider_list ||= []
    end

    # from module.rb
    def self.services
      @@services ||= {}
    end

    def self.find_closest_service(entity, service)
      if service == 'outpatient' || service == :outpatient
        service = 'ambulatory'
      end

      # if service is nil or not supproted by simulation, patient goes to default hospital
      return entity.hospital[:ambulatory] if service.nil?

      case service.to_sym
      when :ambulatory
        return entity.hospital[:ambulatory]
      when :inpatient
        return entity.hospital[:inpatient]
      when :emergency
        return entity.hospital[:emergency]
      end
    end

    def service?(service)
      services_provided.include?(service)
    end

    def increment_encounters
      @utilization[:encounters] += 1
    end

    def increment_procedures
      @utilization[:procedures] += 1
    end

    def increment_labs
      @utilization[:labs] += 1
    end

    def increment_prescriptions
      @utilization[:prescriptions] += 1
    end

    def self.clear
      Provider.provider_list.clear
      Provider.services.clear
    end
  end
end

File: src/main/java/org/mitre/synthea/world/Provider.java
package org.mitre.synthea.world;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.atomic.AtomicInteger;

import javax.swing.text.html.parser.Entity;

import org.mitre.synthea.modules.Person;

import com.google.common.collect.HashBasedTable;
import com.google.common.collect.Table;
import com.google.gson.internal.LinkedTreeMap;
import com.vividsolutions.jts.geom.Coordinate;
import com.vividsolutions.jts.geom.GeometryFactory;
import com.vividsolutions.jts.geom.Point;

public class Provider {
	
	public static final String AMBULATORY = "ambulatory";
	public static final String INPATIENT = "inpatient";
	public static final String EMERGENCY = "emergency";
	public static final String ENCOUNTERS = "encounters";
	public static final String PROCEDURES = "procedures";
	public static final String LABS = "labs";
	public static final String PRESCRIPTIONS = "prescriptions";
	
	// ArrayList of all providers imported
	private static ArrayList<Provider> providerList = new ArrayList<Provider>();
	// Hash of services to Providers that provide them
	private static HashMap<String, ArrayList<Provider>> services = new HashMap<String, ArrayList<Provider>>();
	
	private LinkedTreeMap attributes;
	private Point coordinates;
	private ArrayList<String> services_provided;
	private Table<Integer, String, AtomicInteger> utilization; // row: year, column: type, value: count
	
	public Provider(LinkedTreeMap p) {
		LinkedTreeMap properties = (LinkedTreeMap) p.get("properties");
		attributes = properties;
		String resourceID = (String) p.get("resourceID");
		attributes.put("resourceID", resourceID);

		ArrayList<Double> coorList = (ArrayList<Double>) p.get("coordinates");
		Point coor = new GeometryFactory().createPoint(new Coordinate(coorList.get(0), coorList.get(1)));
		coordinates = coor;
		
		services_provided = new ArrayList<String>();
		String[] servicesList = ( (String) properties.get("services_provided") ).split(" ");
		for(String s : servicesList){
			services_provided.add(s);
			// add provider to hash of services
			if (services.containsKey(s)){
				ArrayList<Provider> l = services.get(s);
				l.add(this);
			} else{
				ArrayList<Provider> l = new ArrayList<Provider>();
				l.add(this);
				services.put(s, l);
			}
		}
		
		utilization = HashBasedTable.create();
	}
	
	public static void clear(){
		providerList.clear();
		services.clear();
	}
	
	public String getResourceID(){
		return attributes.get("resourceID").toString();
	}
	
	public LinkedTreeMap getAttributes(){
		return attributes;
	}
	
	public Point getCoordinates(){
		return coordinates;
	}
	
	public boolean hasService(String service){
		return services_provided.contains(service);
	}
	
	public void incrementEncounters(String encounterType, int year)
	{
		increment(year, ENCOUNTERS);
		increment(year, ENCOUNTERS + "-" + encounterType);
	}
	
	public void incrementProcedures(int year)
	{	
		increment(year, PROCEDURES);
	}
	
	// TODO: increment labs when there are reports
	public void incrementLabs(int year)
	{	
		increment(year, LABS);
	}
	
	public void incrementPrescriptions(int year)
	{	
		increment(year, PRESCRIPTIONS);
	}
	
	private void increment(Integer year, String key)
	{
		if (!utilization.contains(year, key))
		{
			utilization.put(year, key, new AtomicInteger(0));
		}
		
		utilization.get(year, key).incrementAndGet();
	}
	
	public Table<Integer, String, AtomicInteger> getUtilization(){
		return utilization;
	}
	
	public Integer getBedCount(){
		if(attributes.containsKey("bed_count")){
			return Integer.parseInt(attributes.get("bed_count").toString());
		} else {
			return null;
		}
	}
	
	public static Provider findClosestService(Person person, String service){
		if( service.equals("outpatient") || service.equals("wellness")){
			service = AMBULATORY;
		}
		switch(service) {
		case AMBULATORY :
			if( person.getAmbulatoryProvider() == null ){
				person.setAmbulatoryProvider();
			}
			return person.getAmbulatoryProvider();
		case INPATIENT :
			if( person.getInpatientProvider() == null ){
				person.setInpatientProvider();
			}
			return person.getInpatientProvider();
		case EMERGENCY :
			if( person.getEmergencyProvider() == null ){
				person.setEmergencyProvider();
			}
			return person.getEmergencyProvider();
		}
		// if service is null or not supported by simulation, patient goes to ambulatory hospital
		return person.getAmbulatoryProvider();
	}
	
	public static HashMap<String, ArrayList<Provider>> getServices(){
		return services;
	}
}